GitUI/CommandsDialogs/BrowseDialog/FormCommitCount.cs

```diff
@@ -28,69 +34,98 @@ private void cbIncludeSubmodules_CheckedChanged(object sender, EventArgs e)
 
         private void FetchData()
         {
-            Loading.Visible = true;
-
-            CommitCount.Text = "";
-            var dict = new Dictionary<string, HashSet<string>>();
-            var (items, _) = CommitCounter.GroupAllCommitsByContributor(Module);
-            if (cbIncludeSubmodules.Checked)
-            {
-                var submodules = Module.GetSubmodulesLocalPaths();
+            _backgroundLoaderTokenSource.Cancel();
+            _backgroundLoaderTokenSource = new CancellationTokenSource();
+            var token = _backgroundLoaderTokenSource.Token;
 
-                foreach (var submoduleName in submodules)
+            ThreadHelper.JoinableTaskFactory.RunAsync(
+                async () =>
                 {
-                    GitModule submodule = Module.GetSubmodule(submoduleName);
-                    if (submodule.IsValidGitWorkingDir())
+                    Loading.Visible = true;
+
+                    CommitCount.Text = "";
+
+                    var includeSubmodules = cbIncludeSubmodules.Checked;
+
+                    await TaskScheduler.Default.SwitchTo(alwaysYield: true);
+
+                    string text = "";
+
+                    var dict = new Dictionary<string, HashSet<string>>();
+                    var (items, _) = CommitCounter.GroupAllCommitsByContributor(Module);
+
+                    if (includeSubmodules)
                     {
-                        var (submoduleItems, _) = CommitCounter.GroupAllCommitsByContributor(submodule);
-                        foreach (var (name, count) in submoduleItems)
+                        var submodules = Module.GetSubmodulesLocalPaths();
```

Comment:
💡 I'm not sure which operations within this method call into **git.exe**, but you should place the following *before* each one:

```
token.ThrowIfCancellationRequested();
```

Reply:
The loop that generates the text message checks for cancellation and returns an empty string. However each iteration calls git a few times. A one-liner will be less noisy.

My understanding is that throwing is 'fine' because `FileAndForget` swallows cancelled operation exceptions.

Reply:
> The loop that generates the text message checks for cancellation and returns an empty string.

The goal with cancellation is to avoid two things:

* Work that is computationally expensive or has side effects
* Work that blocks the progress of other work

Due to the use of lock files in git, we're targeting the latter here (we want to make sure subsequence fetch requests can be processed as efficiently as possible after an earlier one is cancelled).

> My understanding is that throwing is 'fine' because FileAndForget swallows cancelled operation exceptions.

Correct